### PR TITLE
Better handle large numbers from external data files

### LINKED ISSUE #3363: Large numbers in JSON lead to overflow
### Description

Use this typst code:

```
#let bignum = json("big.json")

#bignum
```

with this json file:

```
18446744073709551615
```

(This is $2^{64}-1$)

and see:
wat

### Operating system

Linux

### Typst version

Using `typst-0.9.0`

### PR DESCRIPTION
**Pull request by elegaanz**
When deserializing JSON files, a `u64` may be produced by serde, that will then be converted to a `i64` when using it as a Typst `Value`. However, some number that can fit in a `u64` are too big for a `i64`.

So far, Typst overflowed when doing the conversion, leading to a negative value where a large positive value was expected. With this change, it will now use a floating point number internally to represent this integer. This comes at the cost of losing precision, but the value will be much closer to the expected one.

Other fixes to this issue were:
- a saturating cast. It was not satisfying in my opinion because the value you would get would be very different from the one you would expect, making it useless.
- using a string instead of an actual number. With that solution, you cannot easily do math on your data. It would also be confusing to have a number become of string type when imported in Typst.
- report an error. I considered that it was better to have imprecise data than being completely blocked because of what the JSON file contains.
- adding a parameter to the `json` function to tell whether or large numbers should result in an error, or in imprecisions. In my opinion, this option added quite a lot of complexity for a use case that most people will never encounter.

Fixes #3363

### CODE REVIEW

**Comment by elegaanz** [comment]
I can maybe add a note about this edge case in the documentation of the `json` function too. What do you think?

File: tests/typ/bugs/3363-json-large-number.typ [hunk 1]
```diff
@@ -0,0 +1,7 @@
+// Big numbers (larger than what i64 can store) should be saturating
+// and not overflowing
```
Comment (reknih) [inline tests/typ/bugs/3363-json-large-number.typ:2]: This comment does not apply to the implementation anymore

File: tests/typ/bugs/3363-json-large-number.typ
Comment (reknih) [inline tests/typ/bugs/3363-json-large-number.typ:1]: This test should be `Ref: false` and use the `test` function since we try to avoid checking in reference images for things that need not be tested visually.

**Comment by reknih** [comment]
> I can maybe add a note about this edge case in the documentation of the `json` function too. What do you think?

Sounds good!

File: crates/typst/src/loading/json.rs [hunk 2]
```diff
@@ -9,13 +9,18 @@ use crate::World;
...
+/// The file must contain a valid JSON value, such as object or array. JSON
+/// objects will be converted into Typst dictionaries, and JSON arrays will be
+/// converted into Typst arrays. Strings and booleans will be converted into the
+/// Typst equivalents, `null` will be converted into `{none}`, and numbers will
+/// be converted to floats or integers depending on whether they are whole
+/// numbers.
 ///
-/// The function returns a dictionary or an array, depending on the JSON file.
+/// Be aware that integers larger than 2<sup>63</sup>-1 will be converted to
+/// floating point numbers, which may give an approximative value.
```
[the remark this answers is not in the document]
Comment (reknih) [reply crates/typst/src/loading/json.rs:20]: This is still on topic for the PR, so we can just rename it and broaden the scope a bit

[the remark this answers is not in the document]
Comment (elegaanz) [reply crates/typst/src/loading/json.rs:20]: Okay! For info:
- CBOR and YAML have the same issue as JSON. I will update the docs for these functions as well.
- CSV and XML consider every primitive value to be a string, so this issue is not relevant
- TOML reports `Failed to parse TOML (number too large to fit in target type at line 1 column 10)` which comes from the `toml` crate directly I believe. Ideally this should be made consistent with JSON, CBOR and YAML, but I think it is not such a big issue, and fixing this would probably be a bit more involved and out of the scope of this PR.

[the remark this answers is not in the document]
Comment (reknih) [reply crates/typst/src/loading/json.rs:20]: Upstream TOML bug: https://github.com/toml-rs/toml/issues/705

[the remark this answers is not in the document]
Comment (reknih) [reply crates/typst/src/loading/json.rs:20]: Sounds good
